// pnpm/lib/cmd/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const help_1 = require("./help");
const import_1 = require("./import");
const install_1 = require("./install");
const installTest_1 = require("./installTest");
const link_1 = require("./link");
const list_1 = require("./list");
const outdated_1 = require("./outdated");
const prune_1 = require("./prune");
const publish_1 = require("./publish");
const rebuild_1 = require("./rebuild");
const recursive_1 = require("./recursive");
const root_1 = require("./root");
const run_1 = require("./run");
const server_1 = require("./server");
const store_1 = require("./store");
const uninstall_1 = require("./uninstall");
const unlink_1 = require("./unlink");
const update_1 = require("./update");
exports.default = {
    help: help_1.default,
    import: import_1.default,
    install: install_1.default,
    installTest: installTest_1.default,
    link: link_1.default,
    list: list_1.default,
    outdated: outdated_1.default,
    pack: publish_1.pack,
    prune: prune_1.default,
    publish: publish_1.default,
    rebuild: rebuild_1.default,
    recursive: recursive_1.default,
    restart: run_1.restart,
    root: root_1.default,
    run: run_1.default,
    server: server_1.default,
    start: run_1.start,
    stop: run_1.stop,
    store: store_1.default,
    test: run_1.test,
    uninstall: uninstall_1.default,
    unlink: unlink_1.default,
    update: update_1.default,
};
