// pnpm/lib/cmd/unlink.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const read_importer_manifest_1 = require("@pnpm/read-importer-manifest");
const supi_1 = require("supi");
const createStoreController_1 = require("../createStoreController");
async function default_1(input, opts) {
    const store = await createStoreController_1.default(opts);
    const unlinkOpts = Object.assign(opts, {
        store: store.path,
        storeController: store.ctrl,
    });
    if (!input || !input.length) {
        return supi_1.mutateModules([
            {
                dependencyNames: input,
                manifest: await read_importer_manifest_1.readImporterManifestOnly(opts.prefix),
                mutation: 'unlinkSome',
                prefix: opts.prefix,
            },
        ], unlinkOpts);
    }
    return supi_1.mutateModules([
        {
            manifest: await read_importer_manifest_1.readImporterManifestOnly(opts.prefix),
            mutation: 'unlink',
            prefix: opts.prefix,
        },
    ], unlinkOpts);
}
exports.default = default_1;
