// renovate/dist/platform/bitbucket-server/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// SEE for the reference https://github.com/renovatebot/renovate/blob/c3e9e572b225085448d94aa121c7ec81c14d3955/lib/platform/bitbucket/utils.js
const url_1 = __importDefault(require("url"));
const bb_got_wrapper_1 = __importDefault(require("./bb-got-wrapper"));
// https://docs.atlassian.com/bitbucket-server/rest/6.0.0/bitbucket-rest.html#idp250
const prStateMapping = {
    MERGED: 'merged',
    DECLINED: 'closed',
    OPEN: 'open',
};
function prInfo(pr) {
    return {
        version: pr.version,
        number: pr.id,
        body: pr.description,
        branchName: pr.fromRef.displayId,
        title: pr.title,
        state: prStateMapping[pr.state],
        createdAt: pr.createdDate,
        canRebase: false,
    };
}
exports.prInfo = prInfo;
const addMaxLength = (inputUrl, limit = 100) => {
    const { search, ...parsedUrl } = url_1.default.parse(inputUrl, true); // eslint-disable-line @typescript-eslint/no-unused-vars
    const maxedUrl = url_1.default.format({
        ...parsedUrl,
        query: { ...parsedUrl.query, limit },
    });
    return maxedUrl;
};
async function accumulateValues(reqUrl, method = 'get', options, limit) {
    let accumulator = [];
    let nextUrl = addMaxLength(reqUrl, limit);
    const lowerCaseMethod = method.toLocaleLowerCase();
    while (typeof nextUrl !== 'undefined') {
        // TODO: fix typing
        const { body } = await bb_got_wrapper_1.default[lowerCaseMethod](nextUrl, options);
        accumulator = [...accumulator, ...body.values];
        if (body.isLastPage !== false)
            break;
        const { search, ...parsedUrl } = url_1.default.parse(nextUrl, true); // eslint-disable-line @typescript-eslint/no-unused-vars
        nextUrl = url_1.default.format({
            ...parsedUrl,
            query: {
                ...parsedUrl.query,
                start: body.nextPageStart,
            },
        });
    }
    return accumulator;
}
exports.accumulateValues = accumulateValues;
